Add new modal options to require confirmation before closing

davwheat opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

> For clarity, this PR was originally for adding attrs which disabled modal dismissing via backdrop clicks and Esc keypresses, but was changed after discussion.

I'm aware that Bootstrap Modals is set to get replaced by Micromodal (or another library) in the near future, but implementing this functionality now should (hopefully) mean it gets carried across to the new library. 🤞 

**Changes proposed in this pull request:**
This provides more customisability for modals, allowing extensions (and core) to determine if confirmation should be required before closing the modal.

The current field `isDismissible` was previously incorrectly labelled as only determining if an X button is visible, not stating that it also controlled if a backdrop click or Esc key press would dismiss the modal. (To be fair, I can't think of any situation where you want to remove a close button and keep Esc and backdrop dismissal.)

This PR adds two new static fields: `requireImplicitCloseConfirmation` and `requireExplicitCloseConfirmation`. This allows extensions to set whether confirmation should be needed for implicit actions, such as backdrop clicks and Esc keypresses.

I've been in the middle of completing several modals with forms in some extensions where I begin to click the button, then decide to edit one of the form entries, so drag my cursor off the submit button onto the backdrop, which then closes the modal. I then either have to re-complete the entire modal again, or just don't bother.

Another situation where backdrop dismissals frustrate me is if I'm copying text from the modal. I tend to click and drag off the end of the component (bottom or sides) to quickly copy all the text I need quickly. Releasing the click causes the modal to dismiss.

While extensions should probably find a way around this, or Bootstrap itself should be patched, in the long run, more customisability is better (in my mind), so exposing these options to extensions is a good thing!

---

This also addresses an issue with Esc-based closure. BS Modals currently only detect Esc presses when performed within their content, which requires that they have focused content. Modals without any autofocused content (no inputs, selects, or textareas) cannot be closed via Esc by default.

This PR disables the default detection techniques used by BS Modal and instead implements custom handlers to match behaviour between the confirmation-based and standard closures, and to fix this issue.

https://user-images.githubusercontent.com/7406822/111080808-0d9bc400-84f8-11eb-8c0e-d0e7fac5e2cf.mp4

**Reviewers should focus on:**
Where should `requireImplicitCloseConfirmation` be overridden to `true`? I've thought of the tag selection modal off the top of my head, but please suggest any other locations.

**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.

davwheat committed:
Add attrs to Modal to allow more dismissal customisability

davwheat committed:
Implement new props into Bootstrap modal options

davwheat committed:
Rename `doesBackdropDismiss` -> `doesBackdropClickDismiss`

askvortsov1 reviewed (changes requested):
Looks like some dist changes were accidentially included, could you remove those before we get into review?

davwheat committed:
Allow `isDismissible` to override other two options

davwheat committed:
Fix Bootstrap disabling Esc when backdrop dismissal is disabled

A small "hack" was needed to do this. If backdrop dismissal is off, we add an event listener to watch for an Escape key-press. If we get one, then we hide the modal.

davwheat committed:
Integrate review comments

davwheat committed:
Simplify `isDismissible` override check

askvortsov1 reviewed (changes requested):
Had a chance to test this out. I get where this is coming from with a desire to prevent accidential closings, but not being able to click outside of the modal to close it is problematic IMO. It definitely shouldn't be the default behavior, but I'm not even sure whether we want to make it an option, since it might be confusing to users.

davwheat commented:
Whoops... accidentally closed when I was cleaning up my fork 🙈

davwheat committed:
Implement close confirmation system

davwheat committed:
Add missing translation

davwheat committed:
Fix bug with keyboard dismissal

davwheat committed:
Remove console log

davwheat committed:
Update JSDoc comment
